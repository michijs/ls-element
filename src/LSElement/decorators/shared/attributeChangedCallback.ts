import { LSCustomElement } from '../../types';
import { getAttributeValue } from '../../utils/getAttribute';

export function attributeChangedCallback(self: LSCustomElement, name: string, oldValue, newValue) {
  if (newValue != oldValue) {
    self.componentWillReceiveAttribute?.(name, newValue, oldValue);
    self[name] = getAttributeValue(newValue);
  }
}