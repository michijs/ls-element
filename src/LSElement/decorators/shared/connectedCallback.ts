import type { LSCustomElement } from '../../types';
import { addReflectedAttributes } from '../../properties/addReflectedAttributes';
import { executeFirstRender } from '../../DOM/executeFirstRender';

export function connectedCallback(self: LSCustomElement) {//Children and attributes should be managed after constructor
  addReflectedAttributes(self);
  if (!self.ls.alreadyRendered) {
    self.componentWillMount?.();
    executeFirstRender(self);
    self.ls.alreadyRendered = true;
    self.componentDidMount?.();
  }
}