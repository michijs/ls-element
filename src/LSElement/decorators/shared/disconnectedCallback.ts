import { removeStoredAttributes } from '../../properties/removeStoredAttributes';
import type { LSCustomElement } from '../../types';

export function disconnectedCallback(self: LSCustomElement) {
  if (self.parentNode === null) {//TODO: search a better way to validate if element does not exists anymore
    removeStoredAttributes(self);
    self.componentDidUnmount?.();
  }
}