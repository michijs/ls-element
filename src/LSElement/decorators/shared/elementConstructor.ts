import type { LSCustomElement, ShadowOption, OtherShadowOptions } from '../../types';
import { addStores } from '../../properties/addStores';
import { initStore } from '../../properties/initStore';
import { addAttributes } from '../../properties/addAttributes';
import { addStoredAttributes } from '../../properties/addStoredAttributes';
import { addObservers } from '../../properties/addObservers';

export function elementConstructor(self: LSCustomElement, shadow: ShadowOption, otherShadowOptions: OtherShadowOptions) {
  if (shadow) {
    const shadowMode = shadow;
    const attachedShadow = self.attachShadow({ mode: shadowMode, ...otherShadowOptions });
    if (shadowMode === 'closed') {
      self._shadowRoot = attachedShadow;
    }
  }
  self.ls = self.ls || {};
  initStore(self);
  addStores(self);
  addAttributes(self);
  addStoredAttributes(self);
  addObservers(self);
}