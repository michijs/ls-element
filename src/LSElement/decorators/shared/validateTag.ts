export const validateTag = (tag: string) => {
  if (tag.indexOf('-') <= 0) {
    throw new Error('You need at least 1 dash in the custom element name!');
  }
};
// TODO: ver si vale la pena