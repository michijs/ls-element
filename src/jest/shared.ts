import { MainElement } from './MainElement';

export type Result =
    'create1000Rows'
    | 'replaceAllRows'
    | 'partialUpdate'
    | 'selectRow'
    | 'swapRows'
    | 'removeRow'
    | 'createManyRows'
    | 'appendRowsToLargeTable'
    | 'clearRows'

const create1000Rows = () => {
  document.getElementById('run').click();
};
const add1000Rows = () => {
  document.getElementById('add').click();
};
const create10000Rows = () => {
  document.getElementById('runlots').click();
};
const updateEvery10Rows = () => {
  document.getElementById('update').click();
};
const swapRows = () => {
  document.getElementById('swaprows').click();
};
const getTableBody = () => {
  return Array.from(document.getElementById('tbody').children);
};
const clear = () => {
  document.getElementById('clear').click();
};
const getRowId = (element: Element) => {
  return Number(element.childNodes.item(0).childNodes.item(0).textContent);
};

export function makePerformanceTests(dataController: () => MainElement, selectByIndex: boolean) {
  const results = new Map<Result, number>();
  const saveResult = (key: Result, functionToMeasure: () => void) => {
    const t0 = performance.now();
    functionToMeasure();
    const t1 = performance.now();
    results.set(key, Number(Number(t1 - t0).toFixed(2)));
  };
  it('creates 1000 rows when clicking run', () => {
    saveResult('create1000Rows', create1000Rows);
    expect(getTableBody().length).toEqual(1000);
  });
  it('replaces 1000 rows when clicking run', () => {
    create1000Rows();
    saveResult('replaceAllRows', create1000Rows);
    expect(getTableBody().length).toEqual(1000);
    getTableBody().forEach(element => {
      expect(getRowId(element)).toBeGreaterThan(1000);
    });
  });
  it('update every 10th row 1000 rows on a table with 1000 rows when clicking update', () => {
    create1000Rows();
    saveResult('partialUpdate', updateEvery10Rows);
    expect(getTableBody().length).toEqual(1000);
    getTableBody().forEach((element, index) => {
      if (index % 10 === 0) {
        expect(element.innerHTML.includes('!!!')).toBeTruthy();
      } else {
        expect(element.innerHTML.includes('!!!')).toBeFalsy();
      }
    });
  });
  it('select a row (1000 rows)', () => {
    create1000Rows();
    saveResult('selectRow', () => dataController().select(999));
    expect(getTableBody()[selectByIndex ? 999 : 998].className).toEqual('danger');
  });
  it('swap a row (1000 rows)', () => {
    create1000Rows();
    saveResult('swapRows', swapRows);
    expect(getRowId(getTableBody()[1])).toEqual(999);
    expect(getRowId(getTableBody()[998])).toEqual(2);
  });
  it('remove a row (1000 rows)', () => {
    create1000Rows();
    saveResult('removeRow', () => dataController().delete(999));
    expect(getTableBody().length).toEqual(999);
  });
  it('creates 10000 rows when clicking runlots', () => {
    saveResult('createManyRows', create10000Rows);
    expect(getTableBody().length).toEqual(10000);
  });
  it('append 1000 rows on a large table', () => {
    create10000Rows();
    saveResult('appendRowsToLargeTable', add1000Rows);
    expect(getTableBody().length).toEqual(11000);
  });
  it('clear rows', () => {
    create1000Rows();
    saveResult('clearRows', clear);
    expect(getTableBody().length).toEqual(0);
  });
  return results;
}