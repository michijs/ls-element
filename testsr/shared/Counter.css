:host {
  display: flex;
  flex-direction: row;
}

span {
  min-width: 60px;
  text-align: center;
}
