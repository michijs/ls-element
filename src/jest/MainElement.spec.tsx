import { h, render } from '..';
import Element, { MainElement } from './MainElement';
import { makePerformanceTests, Result } from './shared';

describe('Performance tests - LS-Element', () => {
  let component: MainElement;

  beforeEach(() => {
    component = render(<Element id="test" />);
  });

  it('matches the snapshot on initial render', () => {
    component.rows = component.buildData(3);
    expect(component.children).toMatchSnapshot();
  });

  const getComponent = () => {
    return component;
  };

  const results = makePerformanceTests(getComponent, false);
  afterAll(() => {
    expect(
      new Map<Result, number>([
        ['create1000Rows', 628.04],
        ['replaceAllRows', 1338.53],
        ['partialUpdate', 89.95],
        ['selectRow', 61.77],
        ['swapRows', 64.19],
        ['removeRow', 56.07],
        ['createManyRows', 7013.5],
        ['appendRowsToLargeTable', 1579.74],
        ['clearRows', 67.7]
      ])).toMatchSnapshot('LS-Element 1.2.6');
    expect(results).toMatchSnapshot(new Date().getTime().toString());
  });
});
