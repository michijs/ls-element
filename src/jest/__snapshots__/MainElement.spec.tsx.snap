// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Performance tests - LS-Element clear rows: 1618790535496 1`] = `
Map {
  "create1000Rows" => 519.92,
  "replaceAllRows" => 662.3,
  "partialUpdate" => 95.51,
  "selectRow" => 69.51,
  "swapRows" => 56.73,
  "removeRow" => 54.1,
  "createManyRows" => 2902.63,
  "appendRowsToLargeTable" => 1460.18,
  "clearRows" => 61.41,
}
`;

exports[`Performance tests - LS-Element clear rows: LS-Element 1.2.6 1`] = `
Map {
  "create1000Rows" => 628.04,
  "replaceAllRows" => 1338.53,
  "partialUpdate" => 89.95,
  "selectRow" => 61.77,
  "swapRows" => 64.19,
  "removeRow" => 56.07,
  "createManyRows" => 7013.5,
  "appendRowsToLargeTable" => 1579.74,
  "clearRows" => 67.7,
}
`;

exports[`Performance tests - LS-Element matches the snapshot on initial render 1`] = `
HTMLCollection [
  <div
    class="container"
    id="container"
  >
    <div
      class="jumbotron"
      id="jumbotron"
    >
      <div
        class="row"
        id="1"
      >
        <div
          class="col-md-6"
          id="2"
        >
          <h1
            id="3"
          >
            LS-Element keyed
          </h1>
        </div>
        <div
          class="col-md-6"
          id="4"
        >
          <div
            class="row"
            id="5"
          >
            <div
              class="col-sm-6 smallpad"
              id="6"
            >
              <button
                class="btn btn-primary btn-block"
                id="run"
                type="button"
              >
                Create 1,000 rows
              </button>
            </div>
            <div
              class="col-sm-6 smallpad"
              id="7"
            >
              <button
                class="btn btn-primary btn-block"
                id="runlots"
                type="button"
              >
                Create 10,000 rows
              </button>
            </div>
            <div
              class="col-sm-6 smallpad"
              id="8"
            >
              <button
                class="btn btn-primary btn-block"
                id="add"
                type="button"
              >
                Append 1,000 rows
              </button>
            </div>
            <div
              class="col-sm-6 smallpad"
              id="9"
            >
              <button
                class="btn btn-primary btn-block"
                id="update"
                type="button"
              >
                Update every 10th row
              </button>
            </div>
            <div
              class="col-sm-6 smallpad"
              id="10"
            >
              <button
                class="btn btn-primary btn-block"
                id="clear"
                type="button"
              >
                Clear
              </button>
            </div>
            <div
              class="col-sm-6 smallpad"
              id="11"
            >
              <button
                class="btn btn-primary btn-block"
                id="swaprows"
                type="button"
              >
                Swap Rows
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
    <table
      class="table table-hover table-striped test-data"
      id="table"
    >
      <tbody
        id="tbody"
      >
        <tr
          id="row-1"
        >
          <td
            class="col-md-1"
            id="row-1-1"
          >
            1
          </td>
          <td
            class="col-md-4"
            id="row-1-2"
          >
            <a
              id="row-1-3"
            >
              unsightly brown pizza
            </a>
          </td>
          <td
            class="col-md-1"
            id="row-1-4"
          >
            <a
              id="row-1-5"
            >
              <span
                aria-hidden="true"
                class="glyphicon glyphicon-remove"
                id="row-1-6"
              />
            </a>
          </td>
          <td
            class="col-md-6"
            id="row-1-7"
          />
        </tr>
        <tr
          id="row-2"
        >
          <td
            class="col-md-1"
            id="row-2-1"
          >
            2
          </td>
          <td
            class="col-md-4"
            id="row-2-2"
          >
            <a
              id="row-2-3"
            >
              important green keyboard
            </a>
          </td>
          <td
            class="col-md-1"
            id="row-2-4"
          >
            <a
              id="row-2-5"
            >
              <span
                aria-hidden="true"
                class="glyphicon glyphicon-remove"
                id="row-2-6"
              />
            </a>
          </td>
          <td
            class="col-md-6"
            id="row-2-7"
          />
        </tr>
        <tr
          id="row-3"
        >
          <td
            class="col-md-1"
            id="row-3-1"
          >
            3
          </td>
          <td
            class="col-md-4"
            id="row-3-2"
          >
            <a
              id="row-3-3"
            >
              adorable brown car
            </a>
          </td>
          <td
            class="col-md-1"
            id="row-3-4"
          >
            <a
              id="row-3-5"
            >
              <span
                aria-hidden="true"
                class="glyphicon glyphicon-remove"
                id="row-3-6"
              />
            </a>
          </td>
          <td
            class="col-md-6"
            id="row-3-7"
          />
        </tr>
      </tbody>
    </table>
    <span
      aria-hidden="true"
      class="preloadicon glyphicon glyphicon-remove"
      id="icon"
    />
  </div>,
]
`;
